README.md
# deidentify

A Python library to de-identify medical records with state-of-the-art NLP methods. Pre-trained models for the Dutch language are available.

This repository shares the resources developed in the following paper:

> J. Trienes, D. Trieschnigg, C. Seifert, and D. Hiemstra. Comparing Rule-based, Feature-based and Deep Neural Methods for De-identification of Dutch Medical Records. In: *Proceedings of the 1st ACM WSDM Health Search and Data Mining Workshop (HSDM)*, 2020.

Read more about the work in our [paper](https://arxiv.org/abs/2001.05714) or [blog post](https://medium.com/nedap/de-identification-of-ehr-using-nlp-a270d40fc442).

## Quick Start

### Installation

Create a new virtual environment with an environment manager of your choice. Then, install `deidentify`:

```sh
pip install deidentify
```

We use the spaCy tokenizer. For good compatibility with the pre-trained models, we recommend using the same spaCy version that we used to train the de-identification models.

```sh
pip install -U "spacy<3" https://github.com/explosion/spacy-models/releases/download/nl_core_news_sm-2.3.0/nl_core_news_sm-2.3.0.tar.gz#egg=nl_core_news_sm==2.3.0
```

### Example Usage

The code below shows how to apply a pre-trained de-identification pipeline to an example document. We provide a [list of available models](#pre-trained-models) below.

```py
from deidentify.base import Document
from deidentify.taggers import FlairTagger
from deidentify.tokenizer import TokenizerFactory

# Create some text
text = (
    "Dit is stukje tekst met daarin de naam Jan Jansen. De patient J. Jansen (e: "
    "j.jnsen@email.com, t: 06-12345678) is 64 jaar oud en woonachtig in Utrecht. Hij werd op 10 "
    "oktober door arts Peter de Visser ontslagen van de kliniek van het UMCU."
)

# Wrap text in document
documents = [
    Document(name='doc_01', text=text)
]

# Select downloaded model
model = 'model_bilstmcrf_ons_fast-v0.2.0'

# Instantiate tokenizer
tokenizer = TokenizerFactory().tokenizer(corpus='ons', disable=("tagger", "ner"))

# Load tagger with a downloaded model file and tokenizer
tagger = FlairTagger(model=model, tokenizer=tokenizer, verbose=False)

# Annotate your documents
annotated_docs = tagger.annotate(documents)
```

This completes the annotation stage. Let's inspect the entities that the tagger found:

```py
from pprint import pprint

first_doc = annotated_docs[0]
pprint(first_doc.annotations)
```

This should print the entities of the first document.

```py
[Annotation(text='Jan Jansen', start=39, end=49, tag='Name', doc_id='', ann_id='T0'),
 Annotation(text='J. Jansen', start=62, end=71, tag='Name', doc_id='', ann_id='T1'),
 Annotation(text='j.jnsen@email.com', start=76, end=93, tag='Email', doc_id='', ann_id='T2'),
 Annotation(text='06-12345678', start=98, end=109, tag='Phone_fax', doc_id='', ann_id='T3'),
 Annotation(text='64 jaar', start=114, end=121, tag='Age', doc_id='', ann_id='T4'),
 Annotation(text='Utrecht', start=143, end=150, tag='Address', doc_id='', ann_id='T5'),
 Annotation(text='10 oktober', start=164, end=174, tag='Date', doc_id='', ann_id='T6'),
 Annotation(text='Peter de Visser', start=185, end=200, tag='Name', doc_id='', ann_id='T7'),
 Annotation(text='UMCU', start=234, end=238, tag='Hospital', doc_id='', ann_id='T8')]
```

#### Mask Annotations

Use masking to replace annotations with placeholders. Example: `Jan Jansen -> [NAME]`

```py
from deidentify.util import mask_annotations

masked_doc = mask_annotations(first_doc)
print(masked_doc.text)
```

Which should print:

> Dit is stukje tekst met daarin de naam [NAME]. De patient [NAME] (e: [EMAIL], t: [PHONE_FAX]) is [AGE] oud en woonachtig in [ADDRESS]. Hij werd op [DATE] door arts [NAME] ontslagen van de kliniek van het [HOSPITAL].

#### Replace Annotations with Surrogates [experimental]

Use sorrogate generation to replace annotations with random but realistic alternatives. Example: `Jan Jansen -> Bart Bakker`. The surrogate replacement strategy follows [Stubbs et al. (2015)](https://doi.org/10.1007/978-3-319-23633-9_27).

```py
from deidentify.util import surrogate_annotations

# The surrogate generation process involves some randomness.
# You can set a seed to make the process deterministic.
iter_docs = surrogate_annotations(docs=[first_doc], seed=1)
surrogate_doc = list(iter_docs)[0]
print(surrogate_doc.text)
```

This code should print:

> Dit is stukje tekst met daarin de naam Gijs Hermelink. De patient G. Hermelink (e: n.qvgjj@spqms.com, t: 06-83662585) is 64 jaar oud en woonachtig in Cothen. Hij werd op 28 juni door arts Jullian van Troost ontslagen van de kliniek van het UMCU.

### Available Taggers

There are currently three taggers that you can use:

   * `DeduceTagger`: A wrapper around the DEDUCE tagger by Menger et al. (2018, [code](https://github.com/vmenger/deduce), [paper](https://www.sciencedirect.com/science/article/abs/pii/S0736585316307365))
   * `CRFTagger`: A CRF tagger using the feature set by Liu et al. (2015, [paper](https://www.sciencedirect.com/science/article/pii/S1532046415001197))
   * `FlairTagger`: A wrapper around the Flair [`SequenceTagger`](https://github.com/zalandoresearch/flair/blob/2d6e89bdfe05644b4e5c7e8327f6ecc6b834ec9e/flair/models/sequence_tagger_model.py#L68) allowing the use of neural architectures such as BiLSTM-CRF. The pre-trained models below use contextualized string embeddings by Akbik et al. (2018, [paper](https://www.aclweb.org/anthology/C18-1139/))

All taggers implement the `deidentify.taggers.TextTagger` interface which you can implement to provide your own taggers.

### Tag Set

Use the `TextTagger.tags` to get a list of supported tags. For the `FlairTagger` in above demo this looks as follows:

```py
>>> tagger.tags
['Internal_Location', 'Age', 'Phone_fax', 'Name', 'SSN', 'Hospital', 'Email', 'Initials', 'O',
'Organization_Company', 'ID', 'Profession', 'Care_Institute', 'Other', 'Date', 'URL_IP', 'Address']
```

### Pre-trained Models

We provide a number of pre-trained models for the Dutch language. The models were developed on the Nedap/University of Twente (NUT) dataset. The dataset consists of 1260 documents from three domains of Dutch healthcare: elderly care, mental care and disabled care (note: in the codebase we sometimes also refer to this dataset as `ons`). More information on the design of the dataset can be found in [our paper](https://arxiv.org/abs/2001.05714).


| Name | Tagger | Lang | Dataset | F1* | Precision* | Recall* | Tags |
|------|--------|----------|---------|----|-----------|--------|--------|
| [DEDUCE (Menger et al., 2018)](https://www.sciencedirect.com/science/article/abs/pii/S0736585316307365)** | `DeduceTagger` | NL | NUT | 0.6649 | 0.8192 | 0.5595 | [8 PHI Tags](https://github.com/nedap/deidentify/blob/168ad67aec586263250900faaf5a756d3b8dd6fa/deidentify/methods/deduce/run_deduce.py#L17) |
| [model_crf_ons_tuned-v0.2.0](https://github.com/nedap/deidentify/releases/tag/model_crf_ons_tuned-v0.2.0) | `CRFTagger` | NL | NUT | 0.8511 | 0.9337 | 0.7820 | [15 PHI Tags](https://github.com/nedap/deidentify/releases/tag/model_crf_ons_tuned-v0.2.0) |
| [model_bilstmcrf_ons_fast-v0.2.0](https://github.com/nedap/deidentify/releases/tag/model_bilstmcrf_ons_fast-v0.2.0) | `FlairTagger`  | NL | NUT | 0.8914 | 0.9101 | 0.8735 | [15 PHI Tags](https://github.com/nedap/deidentify/releases/tag/model_bilstmcrf_ons_fast-v0.2.0) |
| [model_bilstmcrf_ons_large-v0.2.0](https://github.com/nedap/deidentify/releases/tag/model_bilstmcrf_ons_large-v0.2.0) | `FlairTagger` | NL | NUT | 0.8990 | 0.9240 | 0.8754 | [15 PHI Tags](https://github.com/nedap/deidentify/releases/tag/model_bilstmcrf_ons_large-v0.2.0) |

*\*All scores are micro-averaged entity-level precision/recall/F1 obtained on the test portion of each dataset. For additional metrics, see the corresponding model release.*

*\*\*DEDUCE was developed on a dataset of psychiatric nursing notes and treatment plans. The numbers reported here were obtained by applying DEDUCE to our NUT dataset. For more information on the development of DEDUCE, see the paper by [Menger et al. (2018)](https://www.sciencedirect.com/science/article/abs/pii/S0736585316307365).*

## Running Experiments and Training Models

If you have your own dataset of annotated documents and you want to train your own models on it, you can take a look at the following guides:

   * [Convert your data into our corpus format](docs/01_data_format.md)
   * [Train and evaluate your own models](docs/02_train_evaluate_models.md)
   * [Logging and pipeline verbosity](docs/06_pipeline_verbosity.md)

If you want more information on the experiments in our paper, have a look here:

   * [NUT annotation guidelines](docs/03_hsdm2020_nut_annotation_guidelines.md)
   * [Surrogate generation procedure](docs/04_hsdm2020_surrogate_generation.md)
   * [Experiments on English corpora: i2b2/UTHealth and nursing notes](docs/05_hsdm2020_english_datasets.md)

### Computational Environment

When you want to run your own experiments, we assume that you clone this code base locally and execute all scripts under `deidentify/` within the following conda environment:

```sh
# Install package dependencies and add local files to the Python path of that environment.
conda env create -f environment.yml
conda activate deidentify && export PYTHONPATH="${PYTHONPATH}:$(pwd)"
```

## Citation

Please cite the following paper when using `deidentify`:

```bibtex
@inproceedings{Trienes:2020:CRF,
  title={Comparing Rule-based, Feature-based and Deep Neural Methods for De-identification of Dutch Medical Records},
  author={Trienes, Jan and Trieschnigg, Dolf and Seifert, Christin and Hiemstra, Djoerd},
  booktitle = {Proceedings of the 1st ACM WSDM Health Search and Data Mining Workshop},
  series = {{HSDM} 2020},
  year = {2020}
}
```

## Contact

If you have any question, please contact Jan Trienes at jan.trienes@gmail.com.


CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/nedap/deidentify/tree/HEAD)

[Full Changelog](https://github.com/nedap/deidentify/compare/v0.7.3...HEAD)

**Closed issues:**

- TypeError: \_\_init\_\_\(\) missing 1 required positional argument: 'text' while loading the model [\#64](https://github.com/nedap/deidentify/issues/64)

## [v0.7.3](https://github.com/nedap/deidentify/tree/v0.7.3) (2022-05-05)

[Full Changelog](https://github.com/nedap/deidentify/compare/v0.7.2...v0.7.3)

**Closed issues:**

- Language of the Evaluator's Tokenizer is set to English when asked for Dutch or French. [\#62](https://github.com/nedap/deidentify/issues/62)
- Paths when running training [\#57](https://github.com/nedap/deidentify/issues/57)

**Merged pull requests:**

- update project dependencies in setup.py [\#65](https://github.com/nedap/deidentify/pull/65) ([osmalpkoras](https://github.com/osmalpkoras))
- Use correct language for tokenizer during evaluation [\#63](https://github.com/nedap/deidentify/pull/63) ([jantrienes](https://github.com/jantrienes))
- Pin pylint version [\#61](https://github.com/nedap/deidentify/pull/61) ([jantrienes](https://github.com/jantrienes))
- Added German language support [\#60](https://github.com/nedap/deidentify/pull/60) ([bbieniek](https://github.com/bbieniek))

## [v0.7.2](https://github.com/nedap/deidentify/tree/v0.7.2) (2021-06-03)

[Full Changelog](https://github.com/nedap/deidentify/compare/v0.7.1...v0.7.2)

**Closed issues:**

- Surrogation output [\#56](https://github.com/nedap/deidentify/issues/56)
- Surrogate Evaluation Script [\#55](https://github.com/nedap/deidentify/issues/55)
- Surrogation Methods  [\#54](https://github.com/nedap/deidentify/issues/54)

**Merged pull requests:**

- Add language parameter to evaluate\_corpus [\#59](https://github.com/nedap/deidentify/pull/59) ([jantrienes](https://github.com/jantrienes))
- Added French language [\#58](https://github.com/nedap/deidentify/pull/58) ([bbieniek](https://github.com/bbieniek))

## [v0.7.1](https://github.com/nedap/deidentify/tree/v0.7.1) (2021-02-15)

[Full Changelog](https://github.com/nedap/deidentify/compare/v0.7.0...v0.7.1)

**Closed issues:**

- Support spaCy 3 [\#51](https://github.com/nedap/deidentify/issues/51)
- Auto exiting from Google colab while training model [\#50](https://github.com/nedap/deidentify/issues/50)

**Merged pull requests:**

- Add Python 3.9 tests to CI pipeline [\#53](https://github.com/nedap/deidentify/pull/53) ([jantrienes](https://github.com/jantrienes))
- Add spacy 3 compatibility [\#52](https://github.com/nedap/deidentify/pull/52) ([jantrienes](https://github.com/jantrienes))

## [v0.7.0](https://github.com/nedap/deidentify/tree/v0.7.0) (2020-12-16)

[Full Changelog](https://github.com/nedap/deidentify/compare/v0.6.1...v0.7.0)

**Merged pull requests:**

- Run integration tests for Python 3.6/3.7/3.8 on CI [\#48](https://github.com/nedap/deidentify/pull/48) ([jantrienes](https://github.com/jantrienes))
- Move pytest-cov and pylint configuration to setup.cfg [\#47](https://github.com/nedap/deidentify/pull/47) ([jantrienes](https://github.com/jantrienes))
- Remove obsolete conditional in CRF tagger [\#46](https://github.com/nedap/deidentify/pull/46) ([jantrienes](https://github.com/jantrienes))
- Add integration tests for FlairTagger and CRFTagger [\#45](https://github.com/nedap/deidentify/pull/45) ([jantrienes](https://github.com/jantrienes))
- Handle invalid model names in model lookup [\#44](https://github.com/nedap/deidentify/pull/44) ([jantrienes](https://github.com/jantrienes))
- Automate model download if not found in cache [\#43](https://github.com/nedap/deidentify/pull/43) ([jantrienes](https://github.com/jantrienes))
- Upgrade model training dependencies [\#42](https://github.com/nedap/deidentify/pull/42) ([jantrienes](https://github.com/jantrienes))
- Return dict for ignored sents in crf.predict\_marginals [\#41](https://github.com/nedap/deidentify/pull/41) ([jantrienes](https://github.com/jantrienes))
- Add fine-tuning and embedding language flags [\#40](https://github.com/nedap/deidentify/pull/40) ([AIessa](https://github.com/AIessa))

## [v0.6.1](https://github.com/nedap/deidentify/tree/v0.6.1) (2020-10-13)

[Full Changelog](https://github.com/nedap/deidentify/compare/v0.6.0...v0.6.1)

**Closed issues:**

- English model and Re-train [\#33](https://github.com/nedap/deidentify/issues/33)

**Merged pull requests:**

- Correctly handle whitespace in BIO to standoff conversion [\#39](https://github.com/nedap/deidentify/pull/39) ([jantrienes](https://github.com/jantrienes))
- Add flag to save final BiLSTM-CRF model when training on a train-subset [\#38](https://github.com/nedap/deidentify/pull/38) ([jantrienes](https://github.com/jantrienes))
- Expand scope of error handling in date parsing [\#37](https://github.com/nedap/deidentify/pull/37) ([jantrienes](https://github.com/jantrienes))
- Escape regex parameters during name replacements [\#36](https://github.com/nedap/deidentify/pull/36) ([jantrienes](https://github.com/jantrienes))
- Handle platform-specific issue with strftime/strptime [\#35](https://github.com/nedap/deidentify/pull/35) ([jantrienes](https://github.com/jantrienes))

## [v0.6.0](https://github.com/nedap/deidentify/tree/v0.6.0) (2020-09-10)

[Full Changelog](https://github.com/nedap/deidentify/compare/v0.5.2...v0.6.0)

**Merged pull requests:**

- Add customizable error handling for PHI replacement [\#34](https://github.com/nedap/deidentify/pull/34) ([jantrienes](https://github.com/jantrienes))

## [v0.5.2](https://github.com/nedap/deidentify/tree/v0.5.2) (2020-09-07)

[Full Changelog](https://github.com/nedap/deidentify/compare/v0.5.1...v0.5.2)

**Merged pull requests:**

- Bundle generator resources with python package via package\_data [\#32](https://github.com/nedap/deidentify/pull/32) ([jantrienes](https://github.com/jantrienes))

## [v0.5.1](https://github.com/nedap/deidentify/tree/v0.5.1) (2020-09-04)


demo.py
from deidentify.base import Document
from deidentify.taggers import FlairTagger
from deidentify.tokenizer import TokenizerFactory

# Create some text
text = (
    "Dit is stukje tekst met daarin de naam Jan Jansen. De patient J. Jansen (e: "
    "j.jnsen@email.com, t: 06-12345678) is 64 jaar oud en woonachtig in Utrecht. Hij werd op 10 "
    "oktober door arts Peter de Visser ontslagen van de kliniek van het UMCU."
)

# Wrap text in document
documents = [
    Document(name='doc_01', text=text)
]

# Select downloaded model
model = 'model_bilstmcrf_ons_fast-v0.2.0'

# Instantiate tokenizer
tokenizer = TokenizerFactory().tokenizer(corpus='ons', disable=("tagger", "ner"))

# Load tagger with a downloaded model file and tokenizer
tagger = FlairTagger(model=model, tokenizer=tokenizer, verbose=False)

# Annotate your documents
annotated_docs = tagger.annotate(documents)


from pprint import pprint

first_doc = annotated_docs[0]
pprint(first_doc.annotations)


from deidentify.util import mask_annotations

masked_doc = mask_annotations(first_doc)
print(masked_doc.text)


from deidentify.util import surrogate_annotations

iter_docs = surrogate_annotations(docs=[first_doc], seed=1)
surrogate_doc = list(iter_docs)[0]
print(surrogate_doc.text)
